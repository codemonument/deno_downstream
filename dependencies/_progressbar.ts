// export * from "https://deno.land/x/progress@v1.2.4/mod.ts";
// export { default as ProgressBar } from "https://deno.land/x/progress@v1.2.4/mod.ts";

export * from "./progress-fork/mod.ts";
export { default as ProgressBar } from "./progress-fork/mod.ts";
