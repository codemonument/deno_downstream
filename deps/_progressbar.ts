// update to >1.3.0 when pr is merged
export * from "https://deno.land/x/progress@v1.3.7/mod.ts";
export { default as ProgressBar } from "https://deno.land/x/progress@v1.3.7/mod.ts";
