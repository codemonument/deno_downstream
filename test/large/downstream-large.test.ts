/**
 * This test file uses the large test files.
 * Excluded from the other downstream test file to not have to wait for that long every time.
 */

import { downstream } from "@mod";
import { ProgressBar } from "@deps/_progressbar.ts";
import { assert, assertSnapshot } from "@deps/_testing.std.ts";
import { drainStream } from "@testutils/drainStream.ts";
import { File1GB } from "@testutils/testfiles.ts";
import { File10GB } from "@testutils/testfiles.ts";
import { MemMetrics } from "@testutils/MemMetrics.ts";

Deno.test(`Downloading a 10 GB File should be possible and should not use RAM excessively`, async (tc) => {
  const { progressStream, fileStream } = await downstream(File1GB);
  const progressEvents: string[] = [];
  // const memSamples: number[] = [];
  const memMetrics = new MemMetrics();
  const progressBar = new ProgressBar({
    title: "downloading: ",
    total: 100,
  });

  // needs to be done to not crash deno test
  const drainPromise = drainStream(fileStream);

  for await (const progress of progressStream) {
    progressBar.render(Number.parseFloat(progress));
    progressEvents.push(progress);

    // samples memory usage in megabytes and calculates some memory metrics,
    // like the average memory usage
    memMetrics.sampleAndStore();
    // memSamples.push(memUsage);
  }

  memMetrics.log();

  // await Deno.writeTextFile(
  //   "./memSamples.json",
  //   JSON.stringify(memSamples),
  //   { create: true },
  // );
  // console.log(generateMemGraph(memSamples));

  await drainPromise;
});

Deno.test(`Downloading a 10 GB File should be possible and should not use RAM excessively`, async (tc) => {
  const { progressStream, fileStream } = await downstream(File10GB);
  const progressEvents: string[] = [];
  const memMetrics = new MemMetrics();
  const progressBar = new ProgressBar({
    title: "downloading: ",
    total: 100,
  });

  // needs to be done to not crash deno test
  const drainPromise = drainStream(fileStream);

  for await (const progress of progressStream) {
    progressBar.render(Number.parseFloat(progress));
    progressEvents.push(progress);

    // samples memory usage in megabytes and calculates some memory metrics,
    // like the average memory usage
    memMetrics.sampleAndStore();
  }

  memMetrics.log();

  await drainPromise;
});
