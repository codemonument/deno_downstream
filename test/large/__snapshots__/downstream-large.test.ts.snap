export const snapshot = {};

snapshot[`Demo Test 1`] = `64001`;
